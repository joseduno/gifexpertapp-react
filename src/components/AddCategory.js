import React, { useState } from 'react'
import PropTypes from 'prop-types'

const AddCategory = ({setCategory}) => {

    const [inputValue, setInputValue] = useState('')
    const handleOnChange = e => setInputValue(e.target.value)
    const handleSubmit = e => {
        e.preventDefault()  // Evita que refresque la pagina al presionar enter

        if (inputValue.trim().length > 2) {
            {/* No es necesario mandar las categorias como paramentro al componente, 
            en la funcion del hook (setCategories), por defecto se recibe como primer
            parametro, el ultimo estado de las categorias */}
            setCategory(inputValue)
            setInputValue('')
        }

    }

    return (
        <form onSubmit={handleSubmit}>
            <input 
                type="text"
                value={inputValue}
                onChange={handleOnChange}
                placeholder='Buscar'
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired,  // Para obligar a recibir una funcion como argumento
}

export default AddCategory
