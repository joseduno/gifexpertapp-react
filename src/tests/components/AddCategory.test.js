import '@testing-library/jest-dom'
import { shallow } from "enzyme"
import AddCategory from '../../components/AddCategory'

describe('Pruebas en el componente AddCategory', () => {

    const setCategory = jest.fn()  // Para emular una funcion, facilita saber si es llamada o no
    let wrapper

    beforeEach(()=>{
        jest.clearAllMocks()
        // en cada prueba se reenderiza el componente para dejarlo como nuevo
        // si no se hace, los cambios en los tests se mantienen a medida que se van ejecutando
        wrapper = shallow(<AddCategory setCategory={setCategory}/>)
    })

    test('debe renderizar correctamente el componente', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('debe cambiar la caja de texto', () => {
        const input = wrapper.find('input')
        const value = 'Hola Mundo'
        // el segundo argumento es el evento
        input.simulate('change', {target: {value:value}})  // para simular cambio en el input text
    })

    test('No debe postear la informacion si el input tiene menos de 2 caracteres', () => {
        wrapper.find('form').simulate('submit', {preventDefault(){}})
        expect(setCategory).not.toHaveBeenCalled()
    })

    test('Debe llamar al setCategory y limpiar la caja de texto', () => {
        const input = wrapper.find('input')
        input.simulate('change', {target: {value: 'Hola Mundo'}})
        wrapper.find('form').simulate('submit', {preventDefault(){}})

        expect(setCategory).toHaveBeenCalledTimes(1)

        // expect.any(Function), para asegurarse que el argumento de setCategory 
        // sea una algo espec√≠fico (Function, String, Number, etc)
        expect(setCategory).toHaveBeenCalledWith(expect.any(String))
        
        expect(input.props().value).toBe('')
    })
    
})
