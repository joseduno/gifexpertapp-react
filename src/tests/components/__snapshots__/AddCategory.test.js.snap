// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente AddCategory debe renderizar correctamente el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Buscar"
    type="text"
    value=""
  />
</form>
`;
